relayer/src/link/relay_path.rs

```diff
@@ -748,48 +780,70 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
     }
 
     /// Handles updating the client on the source chain
+    /// Returns the height at which the client update was processed
     fn update_client_src(
         &self,
         dst_chain_height: Height,
         tracking_id: &str,
-    ) -> Result<(), LinkError> {
-        if self
-            .src_chain()
-            .proven_client_consensus(self.src_client_id(), dst_chain_height, Height::zero())
-            .is_ok()
-        {
-            return Ok(());
+    ) -> Result<Height, LinkError> {
+        if let Some(update_height) = Self::update_height(
+            self.src_chain(),
+            self.src_client_id().clone(),
+            dst_chain_height,
+        ) {
+            return Ok(update_height);
+        } else {
+            self.do_update_client_src(dst_chain_height, tracking_id, MAX_RETRIES)
         }
+    }
 
-        let mut src_err_ev = None;
-        for _ in 0..MAX_RETRIES {
-            let src_update = self.build_update_client_on_src(dst_chain_height)?;
-            info!(
-                "sending updateClient to client hosted on source chain for height {}",
-                dst_chain_height,
-            );
-
-            let tm = TrackedMsgs::new(src_update, tracking_id);
-
-            let src_tx_events = self
-                .src_chain()
-                .send_messages_and_wait_commit(tm)
-                .map_err(LinkError::relayer)?;
-            info!("result: {}", PrettyEvents(&src_tx_events));
-
-            src_err_ev = src_tx_events
-                .into_iter()
-                .find(|event| matches!(event, IbcEvent::ChainError(_)));
+    /// Perform actual update_client_src with retries.
+    ///
+    /// Note that the retry is only performed in the case when there
+    /// is a ChainError event. It would return error immediately if
+    /// there are other errors returned from calls such as
+    /// build_update_client_on_src.
+    fn do_update_client_src(
+        &self,
+        dst_chain_height: Height,
+        tracking_id: &str,
+        retries_left: usize,
+    ) -> Result<Height, LinkError> {
+        info!( "sending update_client to client hosted on source chain for height {} (retries left: {})", dst_chain_height, retries_left );
 
-            if src_err_ev.is_none() {
-                return Ok(());
+        let src_update = self.build_update_client_on_src(dst_chain_height)?;
+        let tm = TrackedMsgs::new(src_update, tracking_id);
+        let src_tx_events = self
+            .src_chain()
+            .send_messages_and_wait_commit(tm)
+            .map_err(LinkError::relayer)?;
+        info!("result: {}", PrettyEvents(&src_tx_events));
+
+        for event in src_tx_events {
+            match event {
+                err_event @ IbcEvent::ChainError(_) => {
```

Comment:
Also, we're first matching against the chain error here which may have been emitted by a different message. Maybe it's better to do this in two passes - first do a `src_tx_events.iter().find()` for `UpdateClient` and then do another one for `ChainError`.

Reply:
Looks like the order is correct because we want to retry until all `MsgUpdateClient`s succeed, but it would be nice to retry only the ones that failed. So I think it will be better to not return on the first `UpdateClient`. Also, I don't think we're interested in the `UpdateClientEvent` height here, we actually need the height at which this event was emitted. 
There is one other case that isn't handled here which is when there is a `ClientMisbehaviourEvent`. 
I'll try to fix these issues in the next commits.